### PR TITLE
[stdlib] Change _withUnsafeGuaranteedRef to use Builtin.convertUnownedUnsafeToGuaranteed.

### PR DESCRIPTION
**Pull request by gottesmm**
This builtin (which lowers to raw SIL that doesn't use an actual builtin
instruction) allows us to access an unmanaged value at +0 with a language
guarantee rather than relying on the optimizer.

Previously, we did not do this directly since without OSSA, we were scared that
the frontend/optimizer would not be able to safely emit this code. Now that we
have ownership ssa, we are able to ensure that the frontend always copies the +0
value passed into the closure if the value +0 escapes from the closure (either
via a return, storing into memory, or by passing off as a +1 parameter to a
function).

rdar://59735604

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by aschwaighofer** [comment]
As a follow-up we should remove UnsafeGuaranteedPeephole.cpp

**Comment by gottesmm** [comment]
@aschwaighofer SGTM

**Review by lorentey — APPROVED** [review]
LGTM. Is there a way to add a test to catch future regressions, like the one that broke the original implementation?

**Comment by lorentey** [comment]
cc @aschwaighofer

File: stdlib/public/core/Unmanaged.swift [hunk 1]
```diff
@@ -206,10 +206,10 @@ public struct Unmanaged<Instance: AnyObject> {
   public func _withUnsafeGuaranteedRef<Result>(
     _ body: (Instance) throws -> Result
   ) rethrows -> Result {
-    let (guaranteedInstance, token) = Builtin.unsafeGuaranteed(_value)
-    let result = try body(guaranteedInstance)
-    Builtin.unsafeGuaranteedEnd(token)
-    return result
+    var tmp = self
+    let fakeBase: Int? = nil
```
Comment (atrick) [inline stdlib/public/core/Unmanaged.swift]: I think you should explain what `fakeBase` is for, and why a "real" base isn't actually required in this case.

Comment (gottesmm) [reply stdlib/public/core/Unmanaged.swift]: The idea is that the user is stating with this API that they are putting a _fixLifetime around as per the documentation.

Comment (gottesmm) [reply stdlib/public/core/Unmanaged.swift]: @atrick I am happy to do that in a follow on commit.

**Comment by gottesmm** [comment]
@lorentey best way is by adding a benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke test linux platform

**Comment by gottesmm** [comment]
linux failure was in sourcekit-lsp everything else passed.

**Comment by gottesmm** [comment]
@swift-ci smoke test linux platform

**Comment by gottesmm** [comment]
@lorentey I have the perfect benchmark for this actually already. I have a version of Prims that uses UnsafeUnowned directly using Unmanaged, but instead of using _wUGR, it uses takeUnretainedValue I will add a version that uses this API instead and land it before this change.

**Comment by lorentey** [comment]
@gottesmm Ah, this reminded me I still have a benchmark somewhere with three binary tree implementations (vanilla ADT with `@indirect case`, UnsafeMutablePointer, Unmanaged) that exercises this

**Comment by lorentey** [comment]
(But what I meant is a unit test that simply checks for unexpected retain/release calls, so that we get an immediate and clear signal. Benchmarks don't run on every merge...)

**Comment by gottesmm** [comment]
Ok. I’ll do it in a follow on commit. Clack needs this.

**Comment by gottesmm** [comment]
Or nm. Won’t let me do it on my phone.

**Comment by gottesmm** [comment]
I’m going to put in the test and retest

**Comment by gottesmm** [comment]
@lorentey The benchmark for Prims is here: #30049

**Comment by gottesmm** [comment]
I am just going to use Karoy's benchmark that just landed to show the impact here. Then I am going to merge and fix the other issues people brought up in a follow on commit.

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci smoke benchmark

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
Finally, swift-ci does my bidding!

**Comment by gottesmm** [comment]
I'm going to merge this now. The benchmarks will show up when they do. Karoy says locally he validated the perf.

**Pull request merged**